website/blog/2020-02-24-presto-lambda.md

```diff
@@ -0,0 +1,193 @@
+    1. Push the `MethodHandle` object represents `v -> v * captured_factor` on the stack.
+    2. Push `captured_factor` on the stack.
+    3. Invoke `MethodHandle#bindTo` to get a `BoundMethodHandle` represents captured lambda on the top of stack. 
+3. Invoke `transform`. 
+
+Unfortunately, this implementation causes Java to generate a separate customized LambdaForm class for every `MethodHandle#bindTo` call (i.e. per each row). Such excessive runtime class generation quickly fills in metaspace full and causes full GC, see [Presto#7935](https://github.com/prestodb/presto/issues/7935) for reproduction and details. JDK developers also confirms that it is expected to have each `BoundMethodHandle` to be customized independently, and `MethodHandle#bindTo` [is not a good fit for implementing lambda capturing](http://mail.openjdk.java.net/pipermail/mlvm-dev/2017-May/006755.html).
+
+To fix this, we redesign Presto lambda execution via [Presto #8031](https://github.com/prestodb/presto/pull/8031). The key observations are:
+- Lambda capture has to be performed per invocation, as different value will be captured for each row).
+- However, we should use the same class representing captured lambda for every `BIND` call, otherwise we will generate too many classes.
+
+We use the same way as Java handles lambda and capture [3, 4]:
+- Lambda is represented as an object whose type is an functional interface (a.k.a. Single Abstract Method class)
+- The `invokedynamic` instruction is used to perform lambda capture:
+    * During the first `invokedynamic` call, the class represents captured lambda (which is a functional interface) and the method that performs capture will be generated. This step is also called linkage and will only be done once.
+    * Every `invokedynamic` call performs capture and returns an instance of the desired functional interface. 
+
+With this design, `BIND` function will always be fused together with the lambda generation step to generate captured lambda in one single step -- we cannot first generate an object that represents uncaptured lambda, and performs a separate partial application step. Note this implementation also doesn't allow more general higher-order function that returns a function as result.
+
+## Lambda in Aggregation 
+
+While lambda is originally introduced to help query nested data with scalar functions, we also noted it can be used in aggregation functions to allow more flexible analytics. The initial preliminary support of lambda in aggregation is added in [Presto#12084](https://github.com/prestodb/presto/pull/12084), with a [reduce_agg](https://prestodb.io/docs/current/functions/aggregate.html#reduce_agg) function for demonstration purpose. By making the input and combine function to be specified with lambda, this function *conceptually* allows to write User-Defined Aggregation Function (UDAF). The following example shows how to use `reduce_agg` to compute group-wise product (instead of sum):
```

Comment:
While lambda was originally introduced to help query nested data with scalar functions, we also noted it can be used in aggregation functions to allow more flexible analytics. The initial support for lambda in aggregation was added in [Presto#12084](https://github.com/prestodb/presto/pull/12084), with a [reduce_agg](https://prestodb.io/docs/current/functions/aggregate.html#reduce_agg) function for demonstration purpose. `reduce_agg1` *conceputally* allows the creation of User-Defined Aggregation Function (UDAF) by making the input and combine functions lambdas. The following example shows how to use `reduce_agg` to compute group-wise product (instead of sum):